spec/01-unit/01-db/07-db.lua

```diff
@@ -0,0 +1,99 @@
+local mocker = require("spec.fixtures.mocker")
+
+
+local function setup_it_block()
+  mocker.setup(finally, {
+    modules = {
+      {"kong.db.strategies", {
+        new = function()
+          local connector = {
+            infos = function()
+              return {}
+            end,
+            connect_migrations = function()
+              return true
+            end,
+            schema_migrations = function()
+              return {}
+            end,
+            is_014 = function()
+              return { is_014 = false }
+            end,
+            close = function()
+            end,
+          }
+          local strategies = mocker.table_where_every_key_returns({})
+          return connector, strategies
+        end,
+      }},
+      {"kong.db", {}},
+    }
+  })
+end
+
+
+describe("DB", function()
+
+  describe("schema_state", function()
+
+    it("returns the state of migrations", function()
+      setup_it_block()
+
+      local DB = require("kong.db")
+
+      local kong_config = {
+        loaded_plugins = {},
+      }
+      local db, err = DB.new(kong_config, "mock")
+      assert.is_nil(err)
+      assert.is_table(db)
+
+      local state = db:schema_state()
+      assert.is_table(state)
+    end)
+
+  end)
+
+  describe("last_schema_state", function()
+
+    it("returns the last fetched state of migrations", function()
+      setup_it_block()
+
+      local DB = require("kong.db")
+
+      local kong_config = {
+        loaded_plugins = {},
+      }
+      local db, err = DB.new(kong_config, "mock")
+      assert.is_nil(err)
+      assert.is_table(db)
+
+      local state = db:schema_state()
+      assert.is_table(state)
+
+      local last_state = db:last_schema_state()
+
+      -- calling last_schema_state returns the same object
+      assert(state == last_state)
+
+      local last_state_2 = db:last_schema_state()
+
+      -- calling it again returns the same object
+      assert(state == last_state_2)
+
+      local state_2 = db:schema_state()
+      assert.is_table(state_2)
+
+      -- schema_state always returns a new object
+      assert(state ~= state_2)
+
+      local last_state_3 = db:last_schema_state()
+
+      -- the latest object created by schema_state is the one cached
+      assert(state_2 == last_state_3)
```

Comment:
`assert.equal` is preferred since it produces a graceful busted failure, while this assertion produces an error. Failed and error-ed tests are counted differently in busted.

Reply:
No, this produces as graceful failure as well (I just checked by changing to `~=`. I wanted to be explicit I'm checking for pointer equality.

Reply:
`equal` does pointer equality, the `same` assertion only compares contents.

In this case a failed assertion will just give an 'assertion failed' error, where the `equal` assertion would provide feedback of the actual contents of the variables.

Reply:
> feedback of the actual contents of the variables.

And that feedback is misleading, because this test is not about their contents. In the case of two tables that are different objects but structurally equivalent (which is the case I'm trying to catch), it results in a message like

```
spec/01-unit/01-db/07-db.lua:93: Expected objects to be equal.
Passed in:
(table) {
  [legacy_is_014] = false
  [needs_bootstrap] = false
  [new_migrations] = {
    [1] = {
      [migrations] = { ... more }
      [namespace] = 'kong.db.migrations.core'
      [subsystem] = 'core' } } }
Expected:
(table) {
  [legacy_is_014] = false
  [needs_bootstrap] = false
  [new_migrations] = {
    [1] = {
      [migrations] = { ... more }
      [namespace] = 'kong.db.migrations.core'
      [subsystem] = 'core' } } }

stack traceback:
	spec/01-unit/01-db/07-db.lua:93: in function <spec/01-unit/01-db/07-db.lua:59>
```

If one reads the message above, what will they assume? First, they will assume the some deep field that's hidden under `... more` is different. If the tables _are_ structurally different, then one might assume that making them strucuturally identical would fix it, because it's the structures being shown and not the raw pointers, which is not the case.

Compare that to getting an `assertion failed`, going straight to the offending line in the test file and seeing that tables are being compared with `==` (plus, these assertions have comments).

I knew about `assert.equal` and `assert.same`, my choice for `==` was deliberate.

But sure, I can change the comments into assertion messages. :)

Reply:
All opinions here. Best solution would be to change the assertion message and including the table-id in the `equal` assertion in Luassert.